### PR TITLE
Fix exception when decoded value is too large by doubling buffer

### PR DESCRIPTION
**Pull request by karamanolev**
1. No need for the test files. There's a much simpler reproduction, which I included in the tests.
1. Removed the test for NoDelta found. This will be covered in the logic for doubling the buffer. This changes the API a bit, since now you are free to get a larger delta than the second argument. However, I think that this is, in fact, the correct behavior from an API standpoint. Imagine you write your code, test it with reasonable inputs, everything works. One day your production code breaks, because of some edge case. I think cases of larger delta than the input should be handled by the client code, if it's actually an issue. Otherwise you will just surprise them badly.
1. The above actually removes the need for the NoDelta exception, so it's removed as well.
1. `sizeof(int) - 4` is horrible, but I'm too lazy ATM to implement the correct solution, which is to check that you didn't overflow the int when left shifting in those 2 places. The error handling logic will be more complicated as well.
1. The sprintf had a buffer overflow vulnerability. Fixed by using snprintf. I agree this can get a string truncated, but is much better than it was before.

### CODE REVIEW

File: xdelta3/_xdelta3.c [hunk 1]
```diff
@@ -21,13 +21,13 @@ static PyObject * xdelta3_execute(PyObject *self, PyObject *args)
...
     // if the output would be longer than the input itself, there's no point using delta encoding
-    output_alloc = input_size;
+    output_alloc = input_size << buffer_double_cnt;
     output_buf = main_malloc(output_alloc);
     result = xd3_encode_memory(input_bytes, input_size, source_bytes, source_size,
         output_buf, &output_size, output_alloc, flags);
   } else {
     // output shouldn't be bigger than the original plus the delta, but give a little leeway
-    output_alloc = input_size + source_size * 11 / 10;
+    output_alloc = (input_size + source_size * 11 / 10) << buffer_double_cnt;
```
Comment (samuelcolvin) [inline xdelta3/_xdelta3.c:30]: can you remove comments that no longer make sense.

File: xdelta3/_xdelta3.c [hunk 2]
```diff
@@ -21,13 +21,13 @@ static PyObject * xdelta3_execute(PyObject *self, PyObject *args)
 
   if (action == 0) {
     // if the output would be longer than the input itself, there's no point using delta encoding
-    output_alloc = input_size;
+    output_alloc = input_size << buffer_double_cnt;
```
Comment (samuelcolvin) [inline xdelta3/_xdelta3.c:24]: I'm afraid I don't agree with this. Sorry.

If I'm getting a delta that is bigger than the input I want to get an error which I have to deal with as soon as possible. Silently "failing" to produce a useful delta is not good.

Please remove this.

File: tests/test_main.py [hunk 3]
```diff
@@ -57,6 +51,14 @@ def test_version():
     xdelta3.print_version()
 
 
+def test_huge_output():
```
Comment (samuelcolvin) [inline tests/test_main.py:54]: Makes much more sense :smile:

File: xdelta3/_xdelta3.c [hunk 4]
```diff
@@ -39,23 +39,31 @@ static PyObject * xdelta3_execute(PyObject *self, PyObject *args)
...
 
-  if(result == ENOSPC) {
-    if (action == 0) {
-      // all is well, just not efficient delta could be found
-      PyErr_SetString(NoDeltaFound, "No delta found shorter than the input value");
-    } else {
+  if (result == ENOSPC) {
+    // Leave some leeway before we overflow int. This is a bad way, the better way would be to
+    // check above when we left shift.
+    if (buffer_double_cnt < sizeof(int) - 4) {
```
Comment (samuelcolvin) [inline xdelta3/_xdelta3.c:45]: can you at least add a comment explaining what this is doing please.

File: xdelta3/_xdelta3.c [hunk 5]
```diff
@@ -39,23 +39,31 @@ static PyObject * xdelta3_execute(PyObject *self, PyObject *args)
...
       PyErr_SetString(XDeltaError, "Output of decoding delta longer than expected");
+    } else {
+      main_free(output_buf);
+      return _xdelta3_execute(self, args, buffer_double_cnt + 1);
     }
   } else {
     char exc_str[80];
-    sprintf(exc_str, "Error occur executing xdelta3: %s", xd3_strerror(result));
+    snprintf(exc_str, sizeof(exc_str) / sizeof(exc_str[0]),
```
Comment (samuelcolvin) [inline xdelta3/_xdelta3.c:53]: makes sense.

**Review by samuelcolvin — CHANGES_REQUESTED** [review]
sorry for slow response.

Overall looks good, mostly trivial things to change except the behaviour when the delta is bigger than the input where I disagree with you.

**Comment by samuelcolvin** [comment]
@karamanolev do you want to complete this?

**Comment by karamanolev** [comment]
@samuelcolvin I'll address these soon, hopefully I'll find some time. Sorry.

**Comment by samuelcolvin** [comment]
No problems, thank you.
